Update Samples section of contributing.md

BillWagner opened the pull request:
This PR updates the section on creating samples.

This is currently a work in progress.

Note for reviewers @mairaw @richlander 

I've made a recommendation to move the C# tutorials under the /samples/getting-started directory, with the goal being that all tutorials are in /samples/getting-started.  I'd like to get your thoughts on that decision before I move the files and update all the links.

Other work that I'll be doing on this tonight:

. Update the opening paragraphs to provide more guidance on the scope of samples. 

. Recommend that (almost) all new topics include a sample.

. Add some specific coding guidelines for samples that are designed to help readers learn from the sample code.

BillWagner committed:
First draft at the samples section for contributions

Still need to move some files, or update locations in contributing.

svick commented on CONTRIBUTING.md:
```diff
@@ -89,6 +89,36 @@ After build completes, you can preview the resulting site locally using built-in
+Each sample that you create should contain a `readme.md` file. This file should
+contain a short description of the sample (one or two paragraphs.) It should tell
+readers if they want to explore the sample more fully. The readme.md file should also contain
+a link to the live document on the [.net documentation site](http://docs.microsoft.com/dotnet/articles/welcome).
```
> It should tell readers if they want to explore the sample more fully.

I don't understand what is this sentence trying to say.

BillWagner committed:
Re-write awkward sentence

tdykstra commented on CONTRIBUTING.md:
```diff
@@ -89,6 +89,36 @@ After build completes, you can preview the resulting site locally using built-in
+
+- `linq` contains all the 101 LINQ samples that were originally created when LINQ was developed. These don't have topics associated with them yet.
+
+NOTE to Reviewers: The C# Tutorials (/docs/csharp/tutorials) have their sample code in /samples/csharp. I think that
```
"contains all C#" or "contains all Core" statements are misleading. A given sample could fit in all of the first three categories; could give guidance on how to pick which one applies:

1. getting started -- all getting started for core or csharp
2. core-projects -- main purpose of sample is to demo .NET core features
3. csharp -- main purpose of sample is to demo csharp features

mairaw commented on docs/core/testing/unit-testing-with-dotnet-test.md:
````diff
@@ -127,6 +127,7 @@ get xUnit.net to work with .NET Core RTM:
   }
 ```
 
+<<<<<<< 69b9dd581cc765e4865c776fc0fd5b8ae33e5d5c
````
merge issues?

BillWagner replied:
Yup. Fixed and rebased.

BillWagner committed:
fix merge conflict issue.

BillWagner committed:
Fix a couple items in the test topic that were missed.

I think this was a merge conflict. The dotnet new command does now
create libraries, and eveyrthing should be fine.

BillWagner committed:
Update for CI System recommendations

Respond to feedback on where samples should exist.
Add description of which runtimes you should use so that your samples
build on our CI server.

mairaw commented on CONTRIBUTING.md:
```diff
@@ -89,6 +89,42 @@ After build completes, you can preview the resulting site locally using built-in
+Each sample that you create should contain a `readme.md` file. This file should
+contain a short description of the sample (one or two paragraphs.) Your readme.md
+should tell readers what they will learn by exploring this sample. The readme.md file should also contain
+a link to the live document on the [.net documentation site](http://docs.microsoft.com/dotnet/articles/welcome).
```
.net -> .NET

mairaw commented on CONTRIBUTING.md:
```diff
@@ -89,6 +89,42 @@ After build completes, you can preview the resulting site locally using built-in
+the topics you are covering in your documents. Ask yourself what readers will learn from reading
+your topic. What will they learn from building and running your sample? 
+
+NOTE to Reviewers: The C# Tutorials (/docs/csharp/tutorials) have their sample code in /samples/csharp. I think that
```
sounds good

BillWagner committed:
update to match .NET branding.

richlander commented on CONTRIBUTING.md:
```diff
@@ -89,6 +89,42 @@ After build completes, you can preview the resulting site locally using built-in
+# Contributing to samples
 
-See the [Samples Readme](https://github.com/dotnet/core-docs/blob/master/samples/README.md).
+We encourage samples for any of our topics. Readers can download and run the samples. They
```
I prefer to avoid ownership-related words, like "our". How about changing the framing a bit ...

> Readers would love to see samples for all topics in .NET documentation. They like to run them and explore the code and experiment with their own changes. Contributions that add new samples are considered high value and are likely to get accepted.

richlander commented on CONTRIBUTING.md:
```diff
@@ -89,6 +89,42 @@ After build completes, you can preview the resulting site locally using built-in
+All samples should be complete applications or libraries. Where the sample creates a library,
+it should include unit tests or an application that lets readers run the code.
+
+The samples are organized in directories that match the directories under
```
This part could be clearer, mostly around file paths. For example, specify the exact location of `getting-started` to avoid any room for confusion. The `docs/*/tutorials` is too confusing (the star part), IMO, for people new to the repo.

richlander replied:
It's outside the context of your PR, but a great opportunity to raise the feedback and fix it.

I'd love to see us link to the Core contributing docs, since they cover some important and complementary content: https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/contributing.md.

> Before adding content, please submit an issue describing your content proposal.

We can just remove "before adding content,"

> Each author should have a link to an online presence for himself/herself.

"to their online presence" ... simplify the gender issue

> Step 1: Open an Issue describing the article you wish to write and how it relates to existing content. Get approval to write your article.

I'd prefer "Get feedback on your proposal." Step 1 (IMO) is less about getting a required ticket to proceed and more about the maintainers not feeling bad if someone skips step 1, invests a bunch of time in an article we opt not to merge. That is the "open source way", IMO.

> Don't forget to submit

3 of the points in the guidance checklist start out as "don't do blah". I suggest we tell folks what we want them to do. "Do blah".

I see: 

> If updating code samples in /samples/, be sure any line number references in your article remain correct

How about:

>  Be sure any line number references in your article remain correct if updating code samples in /samples/.

I see: 

> The documentation is built using docfx and Markdown. It is hosted on the .NET Core website.

How about:

> The documentation is written in [GitHub-flavored Markdown](https://help.github.com/categories/writing-on-github/) and built using [docfx](http://dotnet.github.io/docfx/). The documented is hosted at [docs.microsoft.com](https://docs.microsoft.com/dotnet).

we should include a statement on when docfx will move off of DNX to .NET Core 1.0

> The conceptual documentation is placed in the docs folder in the root of the repo so most of the work will happen there.

What "work"?

> Using the DNX-based docfx tool, you build the resulting

"the resulting" ???

> After build completes

After "the" build ...

> http://localhost:8080/docs/path

Does "path" take you to a specific location or a 404?

mairaw replied:
@richlander I'm actually working on the rest of the guide:
https://github.com/dotnet/core-docs/pull/668

mairaw replied:
@richlander, a couple of comments on this: 
> I'd love to see us link to the Core contributing docs, since they cover some important and complementary content: https://github.com/dotnet/coreclr/blob/master/Documentation/project-docs/contributing.md.

I think that adding another link here with a lot of things that don't apply to us is confusing. We already link to other materials that they need to read. We should keep this simple IMO. So whatever is important that we're missing  should be added here no?

> Each author should have a link to an online presence for himself/herself.

Not applicable anymore

> Step 1: Open an Issue describing the article you wish to write and how it relates to existing content. Get approval to write your article.

Agree. Fixed. 

> The documentation is written in GitHub-flavored Markdown and built using docfx. The documented is hosted at docs.microsoft.com.

Changed.

> we should include a statement on when docfx will move off of DNX to .NET Core 1.0

I'd have to check when this is actually scheduled to happen. 

> The conceptual documentation is placed in the docs folder in the root of the repo so most of the work will happen there.

Not applicable anymore.

> Does "path" take you to a specific location or a 404?

No. I've previously added an example that actually takes you to an actual location. See if that would work for you.

BillWagner committed:
update contributing.md per feedback.

BillWagner committed:
move teleprompter samples and update links.

BillWagner committed:
move linq tutorial and update links.

BillWagner committed:
update webapi client sample

BillWagner committed:
Move the WeatherService sample.

BillWagner committed:
Update guidance on samples directory.

Ready for a final review.

BillWagner commented:
This should be ready to merge now. Can I get a final review?

/cc @tdykstra

tdykstra commented on CONTRIBUTING.md:
```diff
@@ -101,14 +101,46 @@ This starts the local preview on [localhost:8080](http://localhost:8080). You ca
+
+ - `linq` contains all the 101 LINQ samples that were originally created when LINQ was developed. These don't have topics associated with them yet.
+
+   * The `linq` directory contains samples that are specific to understanding the concepts and implementation behind Language Integrated Query (LINQ). 
```
Is it intentional to have two bullets for linq?

BillWagner replied:
yes, each has a quick summary, and more details about that directory. (Admittedly there aren't many more details.)

tdykstra commented on CONTRIBUTING.md:
```diff
@@ -101,14 +101,46 @@ This starts the local preview on [localhost:8080](http://localhost:8080). You ca
+your topic. What will they learn from building and running your sample? 
+
+Each sample that you create should contain a `readme.md` file. This file should
+contain a short description of the sample (one or two paragraphs.) Your readme.md
```
nit - the period should go outside the closing paren

tdykstra commented on docs/core/testing/unit-testing-with-dotnet-test.md:
```diff
@@ -53,29 +53,11 @@ contains the names of your `src` and `test` directories:
-
-You'll also want to remove `program.cs` and replace it with the class for the
-`PrimeService`. To use TDD, you'll create a failing implementation of the
+Rename`Library.cs` and as 'PrimeService.cs`. To use TDD, you'll create a failing implementation of the
```
and as --> as

tdykstra commented on samples/README.md:
```diff
@@ -43,8 +43,12 @@ If you wish to add a code sample:
+    - `win81-x64`
 
-We will eventually have a CI system in place to build these projects.
+We will eventually have a CI system in place to build these projects shortly.
```
if adding "shortly", remove "eventually"

tdykstra commented:
LGTM :shipit:

BillWagner committed:
Respond to feedback.
